pallets/gated-marketplace/src/functions.rs

```diff
@@ -830,6 +845,77 @@ impl<T: Config> Pallet<T> {
 		Ok(())
 	}
 
+	pub fn do_block_user(
+		authority: T::AccountId,
+		marketplace_id: [u8; 32],
+		user: T::AccountId,
+	) -> DispatchResult {
+		// ensure the marketplace exists
+		ensure!(<Marketplaces<T>>::contains_key(marketplace_id), Error::<T>::MarketplaceNotFound);
+		// ensure the origin is authorized to block users
+		Self::is_authorized(authority.clone(), &marketplace_id, Permission::EnlistBlockedUser)?;
+		//ensure the user is not already a participant of the marketplace
+		ensure!(
+			Self::has_role(user.clone(), &marketplace_id, MarketplaceRole::Participant).is_err(),
+			Error::<T>::UserAlreadyParticipant
+		);
+		// if the user is not blocked, block it
+		//ensure the user is not already blocked
+		ensure!(
+			!Self::is_user_blocked(user.clone(), marketplace_id),
+			Error::<T>::UserAlreadyBlocked
+		);
+
+		// insert the user in the blocked list
+		<BlockedUsersByMarketplace<T>>::try_mutate(marketplace_id, |blocked_list| {
+			blocked_list.try_push(user.clone())
+		})
+		.map_err(|_| Error::<T>::ExceedMaxBlockedUsers)?;
+
+		Self::deposit_event(Event::UserBlocked(marketplace_id, user.clone()));
+		Ok(())
+	}
+
+	pub fn do_unblock_user(
+		authority: T::AccountId,
+		marketplace_id: [u8; 32],
+		user: T::AccountId,
+	) -> DispatchResult {
+		// ensure the marketplace exists
+		ensure!(<Marketplaces<T>>::contains_key(marketplace_id), Error::<T>::MarketplaceNotFound);
+		// ensure the origin is authorized to block users
+		Self::is_authorized(authority.clone(), &marketplace_id, Permission::EnlistBlockedUser)?;
+		//ensure the user is not already a participant of the marketplace
+		ensure!(
+			Self::has_role(user.clone(), &marketplace_id, MarketplaceRole::Participant).is_err(),
+			Error::<T>::UserAlreadyParticipant
+		);
+		//ensure the user is blocked
+		ensure!(Self::is_user_blocked(user.clone(), marketplace_id), Error::<T>::UserIsNotBlocked);
+
+		// remove the user from the block list
+		<BlockedUsersByMarketplace<T>>::try_mutate::<_, _, DispatchError, _>(
+			marketplace_id,
+			|blocked_list| {
+				let user_index = blocked_list
+					.iter()
+					.position(|a| *a == user.clone())
+					.ok_or(Error::<T>::UserNotFound)?;
+				blocked_list.remove(user_index);
+				Ok(())
+			},
+		)?;
+		Self::deposit_event(Event::UserUnblocked(marketplace_id, user.clone()));
+		Ok(())
+	}
+
+	fn is_user_blocked(user: T::AccountId, marketplace_id: [u8; 32]) -> bool {
```

Comment:
Instead of using all the set of `iter, position, ok_or` methods, use a simple `contains_key`